### PR TITLE
Adding Kubernetes authentication to hashi_vault lookup

### PR DESCRIPTION
**Pull request by seanmalloy**
##### SUMMARY
Adding Kubernetes authentication to hashi_vault lookup
    
When running in a Kubernetes pod hashi_vault lookup can now authenticate
to HashiCorp Vault using Kubernetes authentication. The JWT is automatically
pulled from k8s pod, so credentials do not need to be supplied to
Ansible for Vault authentication.
    
https://www.vaultproject.io/docs/auth/kubernetes.html
    
The approle auth_method also gains the ability to specify the Vault
mount point using the mount_point configuration option.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
hashi_vault lookup

##### ANSIBLE VERSION
```
ansible 2.8.0.dev0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/sean/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/sean/projects/ansible/lib/ansible
  executable location = /home/sean/projects/ansible/bin/ansible
  python version = 2.7.15 (default, Oct 15 2018, 15:26:09) [GCC 8.2.1 20180801 (Red Hat 8.2.1-2)]
```

##### ADDITIONAL INFORMATION
N/A

### CODE REVIEW

File: :wq
Comment (s-hertel) [inline :wq]: It looks like you have a duplicate file here.

**Comment by seanmalloy** [comment]
@s-hertel thanks for catching that. I removed the duplicate file.

**Comment by seanmalloy** [comment]
It looks like there are no existing unit or integration tests for hashi_vault lookup. Also, the hvac python module must be installed separately from ansible to use hashi_vault lookup, So I believe it would be difficult(impossible?) to write a unit test for this as the automated CI environment would not have the hvac python module.

Let me know if the above is not correct, and I can try to write the first set of unit tests for hashi_vault lookup.

**Comment by Akasurde** [comment]
@seanmalloy Are you still working on this ? Thanks.

**Comment by seanmalloy** [comment]
@Akasurde yes I am willing to continue working on this PR. I was struggling to get a maintainer to review this pull request. I will get the merge conflict resolved and retest it.

Give me a few days.

**Comment by Akasurde** [comment]
@seanmalloy Great. Let me know once you are done with rebase. Thanks.

File: lib/ansible/plugins/lookup/hashi_vault.py
Comment (Akasurde) [inline lib/ansible/plugins/lookup/hashi_vault.py]: This might shadow built-in `format`

File: lib/ansible/plugins/lookup/hashi_vault.py
Comment (Akasurde) [inline lib/ansible/plugins/lookup/hashi_vault.py]: Use [`fetch_url`](https://github.com/ansible/ansible/blob/240d1a6afb43982f16acebef16778d17aab58160/lib/ansible/module_utils/urls.py#L1197) instead of `requests.get`

File: lib/ansible/plugins/lookup/hashi_vault.py
Comment (Akasurde) [inline lib/ansible/plugins/lookup/hashi_vault.py]: Use `with open(` instead of `open(`

**Review by Akasurde — CHANGES_REQUESTED** [review]
Rest LGTM

File: lib/ansible/plugins/lookup/hashi_vault.py
Comment (Akasurde) [inline lib/ansible/plugins/lookup/hashi_vault.py]:
```suggestion
            with open('/var/run/secrets/kubernetes.io/serviceaccount/token') as file:
```

Comment (seanmalloy) [reply lib/ansible/plugins/lookup/hashi_vault.py]: Fixed

File: lib/ansible/plugins/lookup/hashi_vault.py
Comment (seanmalloy) [inline lib/ansible/plugins/lookup/hashi_vault.py]: Using `fetch_url` instead of `requests` is causing some sort of import issue when `getattr` is called below.

File: lib/ansible/plugins/lookup/hashi_vault.py
Comment (seanmalloy) [inline lib/ansible/plugins/lookup/hashi_vault.py]: The call to `getattr` is blowing up with the below error when the `fetch_url` is imported. Prior to using `fetch_url` I was using the `requests` module and everything was working.

Here is the error that `getattr` is throwing now ...

```
'HashiVault' object has no attribute 'module'
```

**Comment by seanmalloy** [comment]
@Akasurde see above inline comments regarding an issue I'm now facing when importing `fetch_url` instead of `requests`. Any suggestions on how to resolve this?

File: lib/ansible/plugins/lookup/hashi_vault.py
Comment (seanmalloy) [reply lib/ansible/plugins/lookup/hashi_vault.py]: Removed support for GCP, so no an issue anymore.

File: lib/ansible/plugins/lookup/hashi_vault.py
Comment (seanmalloy) [reply lib/ansible/plugins/lookup/hashi_vault.py]: Removed GCP auth support

**Comment by seanmalloy** [comment]
I removed support for using Vault GCE auth. The PR now only includes adding Vault Kubernetes auth. I believe the code is good now, but I just need to do one final test to be sure everything is working as expected.

**Comment by seanmalloy** [comment]
@Akasurde I completed testing everything is working as expected. Let me know if any changes are needed.

File: lib/ansible/plugins/lookup/hashi_vault.py
Comment (Akasurde) [inline lib/ansible/plugins/lookup/hashi_vault.py]:
```suggestion
      description: Role name for vault Kubernetes auth.
```

File: lib/ansible/plugins/lookup/hashi_vault.py
Comment (Akasurde) [inline lib/ansible/plugins/lookup/hashi_vault.py:47]:
```suggestion
      version_added: "2.8" 
      env:
```

**Review by Akasurde — COMMENTED** [review]
Sorry for delay. One last change requested. Thanks.

**Comment by seanmalloy** [comment]
@Akasurde I applied your suggested changes and rebased to pull in the latest changes from `devel`. Let me know if anything else is needed. Sorry for the delay.

**Comment by jborean93** [comment]
Ping @jborean93

**Comment by seanmalloy** [comment]
@@jborean93 thanks for discussing this with me at RH Summit. The CI tests are now passing. I think this is ready for a final review now.

File: lib/ansible/plugins/lookup/hashi_vault.py
Comment (jborean93) [inline lib/ansible/plugins/lookup/hashi_vault.py]: Ansible 2.8 feature freeze in this place, this will have to be 2.9, not sure why CI didn't pick it up sorry.

File: lib/ansible/plugins/lookup/hashi_vault.py
Comment (jborean93) [inline lib/ansible/plugins/lookup/hashi_vault.py:264]: Is there a particular reason for this change in behaviour? More just curious because I see the other `auth_*` methods have a default mount_point as well.

File: lib/ansible/plugins/lookup/hashi_vault.py
Comment (jborean93) [inline lib/ansible/plugins/lookup/hashi_vault.py]: I would use a different var name, `file` is a builtin within Python. I also don't know anything about Kubernetes but should this path be hard coded.

**Review by jborean93 — CHANGES_REQUESTED** [review]
Just a few tiny changes, everything else looks great.

File: lib/ansible/plugins/lookup/hashi_vault.py
Comment (seanmalloy) [reply lib/ansible/plugins/lookup/hashi_vault.py]: Fixed.

File: lib/ansible/plugins/lookup/hashi_vault.py
Comment (seanmalloy) [reply lib/ansible/plugins/lookup/hashi_vault.py:264]: In my Vault environment we do not use the default mount point for the AppRole auth back end. This will allow users like me to pass in a non-default mount point.

File: lib/ansible/plugins/lookup/hashi_vault.py
Comment (seanmalloy) [reply lib/ansible/plugins/lookup/hashi_vault.py]: I changed the variable name to `k8s_token_file`.

I believe it is acceptable to hard code `/var/run/secrets/kubernetes.io/serviceaccount/token`. This is the standard service account token path in Kubernetes and OpenShift. I'm not aware of a way to change the path to the token file in Kubernetes.

**Comment by seanmalloy** [comment]
@jborean93 i make the requested changes. Let me know if any additional changes are needed.

I'll need to test this one more time with the code changes I just made. I'll let you know once I confirm it is still working correctly.

**Review by jborean93 — APPROVED** [review]
Looks good, let me know once you have tested the changes and I’ll merge it in.

**Comment by bmillemathias** [comment]
Hi @jborean93 

This PR has never through the commit, 
considering the huge work done, that would be a pity not to commit. 
Do you consider porting this to a collection ? if yes which one ? 

Best.

**Comment by geerlingguy** [comment]
-label k8s

It looks like this plugin was moved into the `community.general` collection: https://github.com/ansible-collections/community.general/blob/main/plugins/lookup/hashi_vault.py

I would suggest closing this PR and opening a new one in that repository.

**Pull request closed without merging**
